ink/crates/phat_rollup_anchor_ink/src/traits/js_rollup_anchor.rs

```diff
@@ -0,0 +1,151 @@
+use crate::traits::rollup_anchor::RollupAnchorError;
+use ink::prelude::vec::Vec;
+use ink::storage::Lazy;
+use openbrush::contracts::access_control::{self, AccessControlError, RoleType};
+use openbrush::traits::Storage;
+use scale::{Decode, Encode};
+
+
+pub const JS_RA_MANAGER_ROLE: RoleType = ink::selector_id!("JS_RA_MANAGER_ROLE");
+
+#[derive(Debug, Eq, PartialEq, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum JsRollupAnchorError {
+    AccessControlError(AccessControlError),
+}
+
+/// convertor from AccessControlError to JsRollupAnchorError
+impl From<AccessControlError> for JsRollupAnchorError {
+    fn from(error: AccessControlError) -> Self {
+        JsRollupAnchorError::AccessControlError(error)
+    }
+}
+
+type CodeHash = [u8; 32];
+
+/// Message sent to provide the data
+/// response pushed in the queue by the offchain rollup and read by the Ink! smart contract
+#[derive(Encode, Decode)]
+#[cfg_attr(
+    feature = "std",
+    derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout)
+)]
+pub enum ResponseMessage {
+    JsResponse {
+        /// hash of js script executed to get the data
+        js_script_hash: CodeHash,
+        /// hash of data in input of js
+        input_hash: CodeHash,
+        /// hash of settings of js
+        settings_hash: CodeHash,
+        /// response value
+        output_value: Vec<u8>,
+    },
+    Error {
+        /// hash of js script
+        js_script_hash: CodeHash,
+        /// input in js
+        input_value: Vec<u8>,
+        /// hash of settings of js
+        settings_hash: CodeHash,
+        /// when an error occurs
+        error: Vec<u8>,
+    },
+}
+
+/// Enum used in the function on_message_received to return the result
+/// INPUT is the request type sent from the smart contract to phat offchain rollup
+/// OUTPUT is the response type receive by the smart contract
+pub enum MessageReceived<INPUT, OUTPUT> {
+    Ok { output: OUTPUT },
+    Error { input: INPUT, error: Vec<u8> },
+}
+
+#[derive(Default, Debug)]
+#[openbrush::storage_item]
+pub struct Data {
+    /// hash of js script executed to query the data
+    js_script_hash: Lazy<CodeHash>,
+    /// hash of settings given in parameter to js runner
+    settings_hash: Lazy<CodeHash>,
+}
+
+#[openbrush::trait_definition]
+pub trait JsRollupAnchor:
```

Comment:
How does it simplify the use of the JS-based rollup anchor? Could you give some exapmle?

Reply:
You can find an example here: https://github.com/GuiGou12358/decentralized_oracle-graph-api-oracle/blob/main/ink/contracts/graph_api_consumer/lib.rs#L215

The `JsRollupAnchor` trait implements the communication logic between the `JS Phat contract` and the `ink! contract consumer`.
It alows to reduce the complexity of the contrat consumer. The developer: 
 - doesn't have to test if the configured script/settings hashes match with the received ones. All is done and tested in the `JsRollupAnchor` trait.
 - can use directly the business struct without care to the struct JsResponse